src/icon.c

```diff
@@ -205,19 +205,120 @@ GdkPixbuf *get_pixbuf_from_icon(const char *iconname)
+        int bits_per_sample;
+        int n_channels;
+
+        g_variant_get(data,
+                      "(iiibii@ay)",
+                      &width,
+                      &height,
+                      &rowstride,
+                      &has_alpha,
+                      &bits_per_sample,
+                      &n_channels,
+                      &data_variant);
+
+        // note: (A+7)/8 rounds up A to the next byte boundary
+        pixelstride = (n_channels * bits_per_sample + 7)/8;
+        len_expected = (height - 1) * rowstride + width * pixelstride;
+        len_actual = g_variant_get_size(data_variant);
+
+        if (len_actual != len_expected) {
+                LOG_W("Expected image data to be of length %" G_GSIZE_FORMAT
+                      " but got a length of %" G_GSIZE_FORMAT,
+                      len_expected,
+                      len_actual);
+                g_variant_unref(data_variant);
+                return NULL;
+        }
+
+        data_pb = (guchar *) g_memdup(g_variant_get_data(data_variant), len_actual);
 
-        pixbuf = gdk_pixbuf_new_from_data(raw_image->data,
+        pixbuf = gdk_pixbuf_new_from_data(data_pb,
                                           GDK_COLORSPACE_RGB,
-                                          raw_image->has_alpha,
-                                          raw_image->bits_per_sample,
-                                          raw_image->width,
-                                          raw_image->height,
-                                          raw_image->rowstride,
-                                          NULL,
-                                          NULL);
+                                          has_alpha,
+                                          bits_per_sample,
+                                          width,
+                                          height,
+                                          rowstride,
+                                          (GdkPixbufDestroyNotify) g_free,
+                                          data_pb);
+        if (!pixbuf) {
+                /* Dear user, I'm sorry, I'd like to give you a more specific
+                 * error message. But sadly, I can't */
+                LOG_W("Cannot serialise raw icon data into pixbuf.");
+                return NULL;
+        }
+
+        /* To calculate a checksum of the current image, we have to remove
+         * all excess spacers, so that our checksummed memory only contains
+         * real data. */
+        size_t data_chk_len = pixelstride * width * height;
+        unsigned char *data_chk = g_malloc(data_chk_len);
+        size_t rowstride_short = pixelstride * width;
+
+        for (int i = 0; i < height; i++) {
+                memcpy(data_chk + (i*rowstride_short),
+                       data_pb  + (i*rowstride),
+                       rowstride_short);
+        }
+
+        *id = g_compute_checksum_for_data(G_CHECKSUM_MD5, data_chk, data_chk_len);
```

Comment:
How about truncating the hash and storing it in a 64 bit int? Faster comparison, slightly bigger chance of collision.
We'll have to also hash the icons that come from the filesystem though.

Reply:
> For that we cache the pixel buffer and introduce a field, which saves
the current icon's id. The icon_id may be a path or a hash.
So you can compare two notifications by their icon_id field regardless
of their icon type by their icon_id field.


https://github.com/dunst-project/dunst/blob/0886f926cc151937c125ddcd31adef15959d6a32/src/notification.h#L50-L54

Reply:
Yes I got that, I proposed changing `icon_id` to a 64 bit int and hashing all icons regardless. The gains would be minimal thinking more about it so not sure how well it'd work out.

Reply:
> How about truncating the hash and storing it in a 64 bit int? Faster comparison

But initially you have to hash the whole image. For making a comparison faster, whcih would occur maybe, this would be a no brainer.

Look, we have to have `appname`, `body` and `summary` the same, before we test for icon equality.

https://github.com/dunst-project/dunst/blob/0886f926cc151937c125ddcd31adef15959d6a32/src/notification.c#L178-L185

 And the `STR_EQ(a->icon_id, b->icon_id)` will occur roughly a single time for a notification.